docs/creating-a-new-v3-ea.md

```diff
@@ -0,0 +1,655 @@
+  customInputValidation, // If applicable, specify custom input validation
+})
+```
+
+### Endpoint Types
+
+Endpoints contain a type parameter that allows specifying all relevant types needed in a single `EndpointTypes` structure. It allows the developer to specify types for the expected adapter request, adapter response, provider request, provider response, and settings structures. An example is shown below.
+
+```typescript
+import { customSettings } from './config'
+
+interface RequestParams {
+  base: string
+  quote: string
+}
+
+interface ResponseSchema {
+  base: string
+  quote: string
+  price: number
+}
+
+interface RequestBody {
+  symbol: string
+  convert: string
+}
+
+type EndpointTypes = {
+  // Expected adapter request structure received by this endpoint
+  Request: {
+    Params: RequestParams
+  }
+  // Expected adapter response structure returned by the endpoint
+  Response: {
+    Data: ResponseSchema
+    Result: number
+  }
+  // The adapter configs including the custom settings
+  // Set to 'SettingsMap' (provided by the framework) if the adapter does not have custom settings
+  CustomSettings: typeof customSettings
+  // Expected data provider request and response structures. Options for REST and Websocket
+  Provider: {
+    RequestBody: RequestBody // Used for HTTP transports. Set to never if API solely uses query parameters.
+    ResponseBody: ResponseSchema // Used for HTTP transports
+    WsMessage: ProviderMessage // Used for Websocket transports
+  }
+}
+```
+
+### Input Parameters
+
+Input parameters define the structure of the request expected by the endpoint. The framework provides helpful fields to customize input parameters along with validations to ensure requests conform to the structure.
+
+```typescript
+const inputParameters = {
+  param1: {
+    aliases: ['param-one'],
+    required: true,
+    type: 'string',
+    description: 'The symbol of the currency to query',
+    options: ['option1', 'option2'], // Enumerated options
+    default: 'option1',
+    dependsOn: ['param2'], // Other inputs this one depends on
+    exclusive: ['param3'], // Other inputs that cannot be present with this one
+  },
+  param2: {
+    required: false,
+    type: 'string',
+    description: 'Required by param1',
+  },
+  param3: {
+    required: false,
+    type: 'string',
+    description: 'Exclusive from param1',
+  },
+} as const
+```
+
+### Cache Key Generator
+
+The `cacheKeyGenerator` method allows the developer to provide custom logic used to generate a unique cache key for each request. The framework's default behavior is to concat the endpoint name with the stringified input parameters such as `test-{"base":"ETH","quote": "USD"}`.
+
+A potential scenario of where a custom cache key generator would be used is if an input parameter is an array or an object. To ensure that a request generates the same unique key every time, the array or object would need to be sorted. An example of this code is shown below.
```

Comment:
I think, not just here but for the entire guide, we should make it clear when extra efforts should be made to make it non-standard (e.g. cache key generation). We should make it clear that developers should stick to the default unless they have to make it custom.

Reply:
Really good point. I'm offering up some of these things as features when I should be treating them as worst case scenarios.